### PR TITLE
WARN: PerformanceWarning for non-pyarrow fallback

### LINKED ISSUE #42613: PyArrow StringDtype / StringArray fallback policy
In https://github.com/pandas-dev/pandas/issues/35169 / https://github.com/pandas-dev/pandas/issues/42597 we discussed the desired behavior of PyArrow-backed StringArray when a certain method is not implemented in `pyarrow.compute`.

For string methods like `str_normalize`, which aren't currently implemented in `pyarrow.compute`, I believe we (silently) cast from `Pyarrow[string]` to an object-dtype ndarray of Python `str` objects at https://github.com/pandas-dev/pandas/blob/edd5af779a9865e32d22bcd93bf00248565335a1/pandas/core/arrays/string_.py#L527-L528. That's going to be slow and more than doubles the memory usage of the array.

These kinds of performance cliffs are difficult for users to debug. I don't think we should do that conversion on behalf of the user. If something isn't implemented yet, then I think we should raise with a message saying they should convert to `string[python]` dtype first.

If we don't want to raise, we could emit a `PerformanceWarning`, similar to what we do for SparseArray when converting to dense.

### PR DESCRIPTION
**Pull request by mroeschke**
- [x] closes #42613 (Replace xxxx with the Github issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: doc/source/whatsnew/v1.5.0.rst [hunk 1]
```diff
@@ -95,6 +95,7 @@ Other enhancements
 - :meth:`pd.concat` now raises when ``levels`` is given but ``keys`` is None (:issue:`46653`)
 - :meth:`pd.concat` now raises when ``levels`` contains duplicate values (:issue:`46653`)
 - Added ``numeric_only`` argument to :meth:`DataFrame.corr`, :meth:`DataFrame.corrwith`, and :meth:`DataFrame.cov` (:issue:`46560`)
+- A :class:`errors.PerformanceWarning` is now thrown when using with ``string[pyarrow]`` dtype with methods that don't dispatch to ``pyarrow.compute`` methods (:issue:`42613`)
```
Comment (jbrockmendel) [inline doc/source/whatsnew/v1.5.0.rst:98]: "when using with" is awkward.  supposed to be "when using XYZ with"?

File: pandas/core/arrays/string_arrow.py [hunk 2]
```diff
@@ -331,6 +333,7 @@ def _maybe_convert_setitem_value(self, value):
 
     def isin(self, values):
         if pa_version_under2p0:
+            fallback_performancewarning(version="2")
```
Comment (jreback) [inline pandas/core/arrays/string_arrow.py:336]: can you check that we are producing this in tests?

**Comment by jreback** [comment]
hmm lots of warnings not being caught
> pandas/tests/strings/test_find_replace.py::test_contains_moar[string[pyarrow]]
  /home/runner/work/pandas/pandas/pandas/tests/strings/test_find_replace.py:182: PerformanceWarning: Falling back on a non-pyarrow code path which may decrease performance. Upgrade to pyarrow >=4 to possibly suppress this warning.
    result = s.str.contains("a")

File: doc/source/whatsnew/v1.5.0.rst (hunk 1 above)
Comment (mroeschke) [reply doc/source/whatsnew/v1.5.0.rst]: Thanks, changed

File: pandas/core/arrays/string_arrow.py (hunk 2 above)
Comment (mroeschke) [reply pandas/core/arrays/string_arrow.py:336]: Testing this now with `tm.assert_produces_warning`

**Pull request merged**